File: packages/soulos-sdk/python/soulos/hybrid.py
"""Hybrid sidecar client for SoulOS kernel."""

from __future__ import annotations

import json
import logging
import os
from pathlib import Path
from typing import Any, Callable, Awaitable

import httpx

logger = logging.getLogger(__name__)

PROBLEM_CONTENT_TYPE = "application/problem+json"


class SoulOSError(Exception):
    """Raised when kernel returns RFC 7807 Problem Details."""

    def __init__(self, code: str, status: int, detail: str, body: dict[str, Any]):
        self.code = code
        self.status = status
        self.detail = detail
        self.body = body
        super().__init__(f"{code}: {detail}")


def _parse_problem(resp: httpx.Response) -> SoulOSError:
    try:
        body = resp.json()
    except json.JSONDecodeError:
        body = {"detail": resp.text}
    code = body.get("code", "UNKNOWN")
    detail = body.get("detail") or body.get("title") or resp.text
    return SoulOSError(code, resp.status_code, detail, body)


class SoulHybridClient:
    """Identity + memory + hybrid prepare/complete with graceful fallback."""

    def __init__(
        self,
        base_url: str | None = None,
        bot_id: str | None = None,
        enabled: bool | None = None,
        timeout: float = 60.0,
        gateway_secret: str | None = None,
        account_id: str | None = None,
        max_retries: int = 2,
    ) -> None:
        self.base_url = (
            base_url or os.getenv("SOULOS_KERNEL_URL", "http://localhost:8000")
        ).rstrip("/")
        self.bot_id = bot_id or os.getenv("SOULOS_BOT_ID", "").strip() or None
        self.gateway_secret = gateway_secret or os.getenv("SOULOS_GATEWAY_SECRET", "").strip() or None
        self.account_id = account_id or os.getenv("SOULOS_ACCOUNT_ID", "").strip() or None
        if enabled is None:
            self.enabled = os.getenv("SOULOS_ENABLED", "1").lower() not in (
                "0",
                "false",
                "no",
            )
        else:
            self.enabled = enabled
        self.timeout = timeout
        self.max_retries = max_retries
        self._client: httpx.AsyncClient | None = None

    def _request_headers(self) -> dict[str, str]:
        headers: dict[str, str] = {}
        if self.gateway_secret:
            headers["X-SoulOS-Gateway-Secret"] = self.gateway_secret
        if self.account_id:
            headers["X-SoulOS-Account-Id"] = self.account_id
        return headers

    async def _get_client(self) -> httpx.AsyncClient:
        if self._client is None:
            self._client = httpx.AsyncClient(timeout=self.timeout, headers=self._request_headers())
        return self._client

    async def close(self) -> None:
        if self._client is not None:
            await self._client.aclose()
            self._client = None

    async def _request(
        self,
        method: str,
        path: str,
        *,
        json_body: dict[str, Any] | None = None,
    ) -> httpx.Response:
        client = await self._get_client()
        url = f"{self.base_url}{path}"
        last_exc: Exception | None = None
        for attempt in range(self.max_retries + 1):
            try:
                resp = await client.request(method, url, json=json_body)
                if resp.status_code >= 500 and attempt < self.max_retries:
                    continue
                if not resp.is_success:
                    ct = resp.headers.get("content-type", "")
                    if PROBLEM_CONTENT_TYPE in ct or resp.status_code >= 400:
                        raise _parse_problem(resp)
                return resp
            except httpx.HTTPError as e:
                last_exc = e
                if attempt >= self.max_retries:
                    raise
        if last_exc:
            raise last_exc
        raise RuntimeError("request failed")

    async def is_ready(self) -> bool:
        if not self.enabled:
            return False
        try:
            resp = await self._request("GET", "/ready")
            if resp.status_code != 200:
                return False
            data = resp.json()
            return data.get("status") == "ok"
        except (httpx.HTTPError, SoulOSError):
            return False

    async def ensure_avatar(
        self,
        external_key: str,
        soul_path_or_dict: str | Path | dict[str, Any],
        runtime_config: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        if isinstance(soul_path_or_dict, (str, Path)):
            path = Path(soul_path_or_dict)
            soul = json.loads(path.read_text(encoding="utf-8"))
        else:
            soul = soul_path_or_dict
        resp = await self._request(
            "POST",
            "/v1/avatars/ensure",
            json_body={
                "external_key": external_key,
                "soul": soul,
                "runtime_config": runtime_config,
            },
        )
        record = resp.json()
        self.bot_id = record.get("id") or record.get("bot_id") or self.bot_id
        return record

    async def prepare_turn(
        self,
        query: str,
        bot_id: str | None = None,
        session_id: str | None = None,
        top_k: int = 5,
    ) -> dict[str, Any] | None:
        if not self.enabled:
            return None
        bid = bot_id or self.bot_id
        if not bid:
            return None
        try:
            resp = await self._request(
                "POST",
                "/hybrid/prepare",
                json_body={
                    "bot_id": bid,
                    "query": query,
                    "top_k": top_k,
                    "session_id": session_id,
                },
            )
            return resp.json()
        except (httpx.HTTPError, SoulOSError) as e:
            logger.warning("SoulOS prepare_turn failed: %s", e)
            return None

    async def complete_turn(
        self,
        summary: str,
        user_message: str | None = None,
        bot_id: str | None = None,
        session_id: str | None = None,
        reflect: bool = True,
        reflect_async: bool = True,
    ) -> dict[str, Any] | None:
        if not self.enabled:
            return None
        bid = bot_id or self.bot_id
        if not bid:
            return None
        try:
            resp = await self._request(
                "POST",
                "/hybrid/complete",
                json_body={
                    "bot_id": bid,
                    "summary": summary,
                    "user_message": user_message,
                    "session_id": session_id,
                    "reflect": reflect,
                    "reflect_async": reflect_async,
                },
            )
            return resp.json()
        except (httpx.HTTPError, SoulOSError) as e:
            logger.warning("SoulOS complete_turn failed: %s", e)
            return None

    async def ingest_memory(
        self,
        content: str,
        bot_id: str | None = None,
        session_id: str | None = None,
    ) -> dict[str, Any] | None:
        """POST /memory/ingest — seed facts (e.g. Phase A handoff notes)."""
        if not self.enabled:
            return None
        bid = bot_id or self.bot_id
        if not bid:
            return None
        try:
            body: dict[str, Any] = {"bot_id": bid, "content": content}
            if session_id is not None:
                body["session_id"] = session_id
            resp = await self._request("POST", "/memory/ingest", json_body=body)
            return resp.json()
        except (httpx.HTTPError, SoulOSError) as e:
            logger.warning("SoulOS ingest_memory failed: %s", e)
            return None

    async def run_turn(
        self,
        query: str,
        generate: Callable[[str, dict[str, Any]], Awaitable[str]],
        *,
        external_key: str | None = None,
        soul: str | Path | dict[str, Any] | None = None,
        session_id: str | None = None,
        top_k: int = 5,
        reflect: bool = True,
    ) -> dict[str, Any]:
        """Optional ensure → prepare → caller generate → complete."""
        if external_key and soul is not None:
            await self.ensure_avatar(external_key, soul)
        prepared = await self.prepare_turn(query, session_id=session_id, top_k=top_k)
        if not prepared:
            raise SoulOSError("PREPARE_FAILED", 0, "prepare_turn returned no context", {})
        system_prompt = prepared["system_prompt"]
        reply = await generate(system_prompt, prepared)
        completed = await self.complete_turn(
            summary=reply[:2000],
            user_message=query,
            session_id=session_id,
            reflect=reflect,
        )
        return {
            "query": query,
            "system_prompt": system_prompt,
            "reply": reply,
            "prepare": prepared,
            "complete": completed,
        }

File: packages/soulos-sdk/ts/src/hybrid.ts
const DEFAULT_CLOUD_URL = "https://api.soulos.dev";

export type SoulHybridClientConfig = {
  baseUrl?: string;
  apiKey?: string;
  botId?: string;
  enabled?: boolean;
  gatewaySecret?: string;
  accountId?: string;
  timeoutMs?: number;
};

export type EnsureAvatarResponse = {
  id: string;
  name: string;
  role: string;
  baseline_msv: Record<string, unknown>;
  current_msv: Record<string, unknown>;
};

export type ContractContext = {
  contract_id?: string;
  expected_step: string;
  missing_slots: string[];
  filled_slots: Record<string, unknown>;
  reject_tokens: string[];
  ui_progress: { step_index: number; step_count: number; label: string };
  allowed_intents: string[];
  prompt_appendix: string;
  turn_version: number;
};

export type HybridPrepareResponse = {
  bot_id: string;
  identity: Record<string, unknown>;
  memories: string[];
  system_prompt: string;
  inner_monologue: string;
  contract_context?: ContractContext;
};

export type HybridCompleteResponse = {
  status: string;
  ingested?: boolean;
  reflect?: string;
  bot_id?: string;
  current_msv?: Record<string, unknown>;
  turn?: {
    step: string;
    slots: Record<string, unknown>;
    advanced: boolean;
    turn_version: number;
  };
};

export function mergeContractIntoSystemPrompt(
  prepareResponse: HybridPrepareResponse | Record<string, unknown>
): string {
  const base = String(prepareResponse.system_prompt ?? "");
  const ctx = prepareResponse.contract_context as ContractContext | undefined;
  const appendix = ctx?.prompt_appendix;
  if (!appendix) return base;
  if (!base) return appendix;
  return `${base.replace(/\s+$/, "")}\n\n${appendix}`;
}


export class SoulHybridClient {
  private readonly baseUrl: string;
  private readonly apiKey?: string;
  private readonly gatewaySecret?: string;
  private readonly accountId?: string;
  private readonly timeoutMs: number;
  public botId?: string;
  public enabled: boolean;

  constructor(config: SoulHybridClientConfig = {}) {
    if (config.baseUrl) {
      this.baseUrl = config.baseUrl.replace(/\/$/, "");
      this.apiKey = config.apiKey;
    } else if (config.apiKey) {
      this.baseUrl = DEFAULT_CLOUD_URL;
      this.apiKey = config.apiKey;
    } else {
      this.baseUrl = "http://localhost:8000";
    }
    this.botId = config.botId;
    this.gatewaySecret = config.gatewaySecret;
    this.accountId = config.accountId;
    this.timeoutMs = config.timeoutMs ?? 60000;
    this.enabled = config.enabled ?? true;
  }

  private headers(): Record<string, string> {
    const h: Record<string, string> = { "Content-Type": "application/json" };
    if (this.apiKey) h["Authorization"] = `Bearer ${this.apiKey}`;
    if (this.gatewaySecret) h["X-SoulOS-Gateway-Secret"] = this.gatewaySecret;
    if (this.accountId) h["X-SoulOS-Account-Id"] = this.accountId;
    return h;
  }

  async isReady(): Promise<boolean> {
    if (!this.enabled) return false;
    try {
      const res = await fetch(`${this.baseUrl}/ready`, { headers: this.headers() });
      if (!res.ok) return false;
      const data = await res.json();
      return data.status === "ok";
    } catch {
      return false;
    }
  }

  async ensureAvatar(
    externalKey: string,
    soul: Record<string, unknown>,
    runtimeConfig?: Record<string, unknown>
  ): Promise<EnsureAvatarResponse> {
    const res = await fetch(`${this.baseUrl}/v1/avatars/ensure`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify({
        external_key: externalKey,
        soul,
        runtime_config: runtimeConfig,
      }),
    });
    const body = await res.json();
    if (!res.ok) {
      throw new Error(body.detail || `ensureAvatar failed (${res.status})`);
    }
    this.botId = body.id;
    return body as EnsureAvatarResponse;
  }

  async prepareTurn(
    query: string,
    options: {
      botId?: string;
      sessionId?: string;
      topK?: number;
    } = {}
  ): Promise<HybridPrepareResponse | null> {
    if (!this.enabled) return null;
    const botId = options.botId ?? this.botId;
    if (!botId) return null;
    try {
      const res = await fetch(`${this.baseUrl}/hybrid/prepare`, {
        method: "POST",
        headers: this.headers(),
        body: JSON.stringify({
          bot_id: botId,
          query,
          session_id: options.sessionId,
          top_k: options.topK ?? 5,
        }),
      });
      if (!res.ok) return null;
      return (await res.json()) as HybridPrepareResponse;
    } catch {
      return null;
    }
  }

  async completeTurn(
    summary: string,
    options: {
      userMessage?: string;
      botId?: string;
      sessionId?: string;
      reflect?: boolean;
      reflectAsync?: boolean;
      filledSlots?: Record<string, unknown> | null;
      intent?: string;
      assistantText?: string;
      expectedVersion?: number;
      idempotencyKey?: string;
      advance?: boolean;
      expectedStep?: string;
      raiseOnError?: boolean;
    } = {}
  ): Promise<HybridCompleteResponse | null> {
    if (!this.enabled) return null;
    const botId = options.botId ?? this.botId;
    if (!botId) return null;
    const reflect = options.reflect ?? Boolean(options.userMessage);
    const contractMode =
      options.filledSlots !== undefined ||
      options.expectedVersion !== undefined ||
      options.intent !== undefined ||
      options.idempotencyKey !== undefined;
    const idempotencyKey =
      options.idempotencyKey ?? (contractMode ? crypto.randomUUID() : undefined);
    const body: Record<string, unknown> = {
      bot_id: botId,
      summary,
      user_message: options.userMessage,
      session_id: options.sessionId,
      reflect,
      reflect_async: options.reflectAsync ?? true,
      advance: options.advance ?? true,
    };
    if (options.filledSlots !== undefined) body.filled_slots = options.filledSlots;
    if (options.intent !== undefined) body.intent = options.intent;
    if (options.assistantText !== undefined) body.assistant_text = options.assistantText;
    if (options.expectedVersion !== undefined) body.expected_version = options.expectedVersion;
    if (idempotencyKey !== undefined) body.idempotency_key = idempotencyKey;
    if (options.expectedStep !== undefined) body.expected_step = options.expectedStep;
    try {
      const res = await fetch(`${this.baseUrl}/hybrid/complete`, {
        method: "POST",
        headers: this.headers(),
        body: JSON.stringify(body),
      });
      if (!res.ok) {
        const problem = await res.json().catch(() => ({}));
        const code = String(problem.code || "");
        if (
          options.raiseOnError ||
          (contractMode && code.startsWith("TURN_"))
        ) {
          const err = new Error(problem.detail || `completeTurn failed (${res.status})`) as Error & {
            code?: string;
            status?: number;
            body?: Record<string, unknown>;
          };
          err.code = code || "UNKNOWN";
          err.status = res.status;
          err.body = problem;
          throw err;
        }
        return null;
      }
      return (await res.json()) as HybridCompleteResponse;
    } catch (e) {
      const code = (e as { code?: string }).code;
      if (
        options.raiseOnError ||
        (typeof code === "string" && code.startsWith("TURN_"))
      ) {
        throw e;
      }
      return null;
    }
  }

  async runTurn(
    query: string,
    generate: (systemPrompt: string, ctx: HybridPrepareResponse) => Promise<string>,
    options: {
      externalKey?: string;
      soul?: Record<string, unknown>;
      sessionId?: string;
      topK?: number;
      reflect?: boolean;
      mergeContract?: boolean;
      filledSlots?: Record<string, unknown> | null;
      intent?: string;
      advance?: boolean;
    } = {}
  ): Promise<{ reply: string; systemPrompt: string; prepare: HybridPrepareResponse; complete: HybridCompleteResponse | null }> {
    if (options.externalKey && options.soul) {
      await this.ensureAvatar(options.externalKey, options.soul);
    }
    const prepared = await this.prepareTurn(query, {
      sessionId: options.sessionId,
      topK: options.topK,
    });
    if (!prepared) {
      throw new Error("prepare_turn returned no context");
    }
    const mergeContract = options.mergeContract !== false;
    const systemPrompt = mergeContract
      ? mergeContractIntoSystemPrompt(prepared)
      : prepared.system_prompt;
    const reply = await generate(systemPrompt, prepared);
    const ctx = prepared.contract_context;
    const completed = await this.completeTurn(reply.slice(0, 2000), {
      userMessage: query,
      sessionId: options.sessionId,
      reflect: options.reflect ?? true,
      assistantText: reply,
      advance: options.advance ?? true,
      raiseOnError: true,
      ...(ctx?.turn_version !== undefined
        ? {
            expectedVersion: ctx.turn_version,
            expectedStep: ctx.expected_step,
          }
        : {}),
      ...(options.filledSlots !== undefined ? { filledSlots: options.filledSlots } : {}),
      ...(options.intent !== undefined ? { intent: options.intent } : {}),
    });
    return {
      reply,
      systemPrompt,
      prepare: prepared,
      complete: completed,
    };
  }
}
